### PR TITLE
MULE-16523: A filter in a transactional scope causes a NullPointerException.

### PR DESCRIPTION
**Pull request by emariotti3**

### CODE REVIEW

File: tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+import org.junit.Test;
+
+public class TransactionExpressionFilterTestCase extends FunctionalTestCase
+{
+
+    protected static final int TIMEOUT = 5000;
```
Comment (fsgonz) [inline tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java]: I think there is a default constant for the response timeout. RECEIVE_TIMEOUT

File: tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    @Override
+    protected String getConfigFile()
+    {
+        return "org/mule/test/integration/transaction/transaction-expression-filter-test-config.xml";
+    }
+
+    @Test
+    public void transactionWithNestedExpressionFilterStopsFlowAndReturnsNull()
+    {
+        List<Integer> payload = Arrays.asList(1, 2, 3);
```
Comment (fsgonz) [inline tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java]: Is there a rationale behind sending a collection as a payload?

File: tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+        MuleClient client = muleContext.getClient();
+        MuleMessage message = new DefaultMuleMessage(payload, mock(Map.class) , muleContext);
+        try
```
Comment (fsgonz) [inline tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java]: Remove this try. Add the throws Exception and let it fail if an unexpected exception is thrown.

File: tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        return "org/mule/test/integration/transaction/transaction-expression-filter-test-config.xml";
+    }
+
+    @Test
+    public void transactionWithNestedExpressionFilterStopsFlowAndReturnsNull()
+    {
+        List<Integer> payload = Arrays.asList(1, 2, 3);
+
+        MuleClient client = muleContext.getClient();
+        MuleMessage message = new DefaultMuleMessage(payload, mock(Map.class) , muleContext);
```
Comment (fsgonz) [inline tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java]: Is the creation of a message really necessary for the test? Can't we just simply send an object (for example an a test string) and let the client create the message for us according to that payload?

File: tests/integration/src/test/resources/org/mule/test/integration/transaction/transaction-expression-filter-test-config.xml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+      xsi:schemaLocation="
+       http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
+       http://www.mulesoft.org/schema/mule/vm http://www.mulesoft.org/schema/mule/vm/current/mule-vm.xsd">
+
+     <vm:connector name="default"/>
+
+     <flow name="expression-in-transaction-flow">
+          <vm:inbound-endpoint path="transaction-filter-all" connector-ref="default"/>
+          <transactional action="ALWAYS_BEGIN">
```
Comment (fsgonz) [inline tests/integration/src/test/resources/org/mule/test/integration/transaction/transaction-expression-filter-test-config.xml]: Does this fail only in the case of a transactional scope?

File: tests/integration/src/test/resources/org/mule/test/integration/transaction/transaction-expression-filter-test-config.xml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<mule xmlns="http://www.mulesoft.org/schema/mule/core"
+      xmlns:vm="http://www.mulesoft.org/schema/mule/vm"
+      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+      xsi:schemaLocation="
+       http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
+       http://www.mulesoft.org/schema/mule/vm http://www.mulesoft.org/schema/mule/vm/current/mule-vm.xsd">
+
+     <vm:connector name="default"/>
```
Comment (fsgonz) [inline tests/integration/src/test/resources/org/mule/test/integration/transaction/transaction-expression-filter-test-config.xml]: Is the definition of the default connector needed?

File: tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java (hunk 2 above)
Comment (emariotti3) [reply tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java]: Not really, this was useful for a previous test version.

File: tests/integration/src/test/resources/org/mule/test/integration/transaction/transaction-expression-filter-test-config.xml (hunk 5 above)
Comment (emariotti3) [reply tests/integration/src/test/resources/org/mule/test/integration/transaction/transaction-expression-filter-test-config.xml]: yes

File: tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java [hunk 7]
```diff
@@ -6,48 +6,42 @@
...
-        {
-            client.dispatch("vm://transaction-filter-all", message);
-            MuleMessage reply = client.request("vm://transaction-filter-all-reply", TIMEOUT);
-            assertNull(reply);
-        }
-        catch (Exception e)
-        {
-            fail(String.format("Exception was thrown on executing transaction with nested expression filter: %s", e.toString()));
-        }
+        client.dispatch("vm://transaction-filter-nothing", "ok", null);
```
Comment (fsgonz) [inline tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java]: Use an already existing dummy message payload like TEST_MESSAGE.

File: tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    @Override
+    protected String getConfigFile()
+    {
+        return "org/mule/test/integration/transaction/transaction-expression-filter-test-config.xml";
+    }
+
+    @Test
+    public void transactionWithNestedExpressionFilterEvaluatesFalseStopsFlowAndReturnsNull() throws Exception
+    {
+        MuleClient client = muleContext.getClient();
```
Comment (fsgonz) [inline tests/integration/src/test/java/org/mule/test/integration/transaction/TransactionExpressionFilterTestCase.java]: Refactor the dispatch logic in a method in order not to duplicate code.

**Pull request merged**